### PR TITLE
Adds filesToOpen to vscode.openFolder (wip)

### LINKED ISSUE #122071: Allow `vscode.openFolder` to optionally take a list of paths that should open in the resulting workspace
I really would like to use this functionality for RemoteHub, for the case where you open to a GitHub blob link. I want the new workspace to be open to the repository, but also open the file.

I have this basically working today, but it is kind of hacky with me having to essentially message pass from the previous workspace to the new one via global storage.

### PR DESCRIPTION
**Pull request by eamodio**
This PR fixes #122071

@bpasero Here is the first shot at this, but it currently doesn't work. 

The first reason is because we are clearing out the files here (I think because we are opening the welcome/getting started/readme):
https://github.com/microsoft/vscode/blob/cdb8d62f088d0678229877895bfc3875ab5e07f8/src/vs/platform/windows/electron-main/windowsMainService.ts#L377

And if I comment that out as a test, the file does get opened, but its opened as a new file with no contents -- seems like its not waiting for the fs providers to start.

### CODE REVIEW

**Comment by bpasero** [comment]
@eamodio when I tried with this code:

```ts
vscode.commands.executeCommand('_files.windowOpen', [
	{ folderUri: vscode.Uri.file('/Users/bpasero/Desktop/test-ts/src/test') },
	{ fileUri: vscode.Uri.file('/Users/bpasero/Development/Microsoft/monaco/src/vs/workbench/browser/actions/workspaceCommands.ts') },
	{ fileUri: vscode.Uri.file('/Users/bpasero/Desktop/test-ts/src/test/runTest.ts') }
]);
```

I was able to open the folder and the files within. Are you saying it is not working with virtual folders and files?

The code in https://github.com/microsoft/vscode/blob/cdb8d62f088d0678229877895bfc3875ab5e07f8/src/vs/platform/windows/electron-main/windowsMainService.ts#L377 only makes sure that once a window has handles the files to open, we clear that variable so that the next window that opens (if there are more) does not open the same files again.

**Comment by eamodio** [comment]
@bpasero Unless I commented out that `filesToOpen = undefined` line, I never saw any files get opened in the re-used window.

When I commented it out, trying with virtual files (`vscode-vfs://`), the file opened with the file name in the tab, but had no contents and had the dirty indicator (like it was just created)

**Comment by bpasero** [comment]
@eamodio took a brief look, and it will need a lot more changes to support to open extension provided files.

First of all, we seem to validate if a file exists to open on startup here and this somehow would need to let files through that have no file system provider yet (?):

https://github.com/microsoft/vscode/blob/b57a9d16c936d6b061ce96a807534906a25768a8/src/vs/workbench/common/editor.ts#L1544-L1544

Even passed that, we will eventually restore the editors here:

https://github.com/microsoft/vscode/blob/e1727abb4114388c7b126f22348b4c4149cf9a2b/src/vs/workbench/browser/layout.ts#L700-L700

But this call does not await extension activation, so it fails. Since this code is on the critical startup path (perf wise), we do not await extensions here. Maybe here we need to await for extensions to activate if the resource scheme is unknown?

**Review by bpasero — CHANGES_REQUESTED** [review]
https://github.com/microsoft/vscode/pull/122268#issuecomment-827392837

File: src/vs/workbench/browser/actions/workspaceCommands.ts [hunk 1]
```diff
@@ -156,14 +161,31 @@ CommandsRegistry.registerCommand({
...
+			}));
+		}
+
 		const options: IOpenWindowOptions = {
 			forceNewWindow: arg?.forceNewWindow,
 			forceReuseWindow: arg?.forceReuseWindow,
 			noRecentEntry: arg?.noRecentEntry,
-			remoteAuthority: arg?.forceLocalWindow ? null : undefined
+			remoteAuthority: arg?.forceLocalWindow ? null : undefined,
+			gotoLineMode: gotoLineMode
```
Comment (bpasero) [inline src/vs/workbench/browser/actions/workspaceCommands.ts:185]:
```suggestion
			gotoLineMode
```

File: src/vs/workbench/browser/actions/workspaceCommands.ts [hunk 2]
```diff
@@ -156,14 +161,31 @@ CommandsRegistry.registerCommand({
...
 		};
-		const uriToOpen: IWindowOpenable = (hasWorkspaceFileExtension(uri) || uri.scheme === Schemas.untitled) ? { workspaceUri: uri } : { folderUri: uri };
-		return commandService.executeCommand('_files.windowOpen', [uriToOpen], options);
+
+		return commandService.executeCommand('_files.windowOpen', urisToOpen, options);
 	},
 	description: {
```
Comment (bpasero) [inline src/vs/workbench/browser/actions/workspaceCommands.ts:190]: The new options need to be documented also here.

File: src/vs/workbench/browser/actions/workspaceCommands.ts [hunk 3]
```diff
@@ -131,6 +131,11 @@ interface IOpenFolderAPICommandOptions {
 	forceReuseWindow?: boolean;
 	noRecentEntry?: boolean;
 	forceLocalWindow?: boolean;
+	filesToOpen?: {
+		uri: URI;
+		line?: number;
```
Comment (bpasero) [inline src/vs/workbench/browser/actions/workspaceCommands.ts:136]: I wonder whether this should align closer to the API we allow for `vscode.open`:

https://github.com/microsoft/vscode/blob/292cc7f4c48fd3f95d6d3709a0d5872ec35a5039/src/vs/workbench/browser/parts/editor/editorCommands.ts#L492-L492

Where we allow a `ITextEditorSelection`:

https://github.com/microsoft/vscode/blob/22cf36faf81ca270390ec233389a81339b40b63a/src/vs/platform/editor/common/editor.ts#L297-L302

Unfortunately, it doesn't look like we support full selection though from where we pass the URI on:

https://github.com/microsoft/vscode/blob/7ac9328da83e00305eb393da74d110b81a1a6204/src/vs/workbench/contrib/files/browser/fileCommands.ts#L94-L94

**Review by bpasero — CHANGES_REQUESTED** [review]
@eamodio I think meanwhile we actually do the proper extension activation, so maybe try this out again:

https://github.com/microsoft/vscode/blob/cc4ace196cf5482f801b2ac06128bace2906f874/src/vs/workbench/common/editor.ts#L1129-L1129

**Pull request closed without merging**